tests/db_tests/test_queries.py

```diff
@@ -481,3 +484,27 @@ def test_project_if_no_outcomes(seeded_test_client, additional_test_data):
 
     test_df = pd.read_sql(test_query_proj.statement, con=db.engine.connect())
     assert list(test_df["project_name"])
+
+
+def test_with_transaction_max_retries(mocker, test_client):
+    """
+    Test the behavior of 'with_transaction' function in case of IntegrityError.
+
+    This test simulates a scenario where a function raises an IntegrityError
+    and ensures that the 'with_transaction' function retries the operation up to the
+    specified maximum number of times. It also checks whether the error is properly
+    logged after reaching the maximum number of retries.
+    """
+    max_retries = 3
+    sleep_duration = 0.1
```

Comment:
You should mock sleep, then assert it is called rather than actually running a sleep method on tests